org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/internal/XtextPluginImages.java

```diff
@@ -139,7 +139,10 @@ public static Image get(String key) {
 	}
 
 	private static void ensureInitialized() {
-		if (PLUGIN_REGISTRY == null)
+		if (PLUGIN_REGISTRY == null &&
```

Comment:
Assuming we come from `get(String)`, wouldn't we immediately see an NPE on `PLUGIN_REGISTRY.get(..)` in line 138?

Reply:
Assuming we come on a non-display thread from `getAnnotationImagesNonfixable` -> `initialize` won't be called and the map is returned on that non-display thread. At the same time the display thread might initialize the map and call `put` on it. This is prone to all sorts of race conditions and exceptions.

I don't feel like this is the correct fix for the ticket.

Reply:
@szarnekow you're right: I had missed `get` and other methods like `org.eclipse.xtext.ui.internal.XtextPluginImages.manage(String, ImageDescriptor)` that simply uses `PLUGIN_REGISTRY`, even without calling `ensureInitialized`. I was assuming that `ensureInitialized` was consistently called. Even so, it is assumed that after the call `PLUGIN_REGISTRY` is not null, which is not the case anymore.

Please let me know if you have other proposals.

Reply:
This class has the final but mutable maps (annotationXyz) as well as the non-final static, lazily initialized PLUGIN_REGISTRY; all of which luckily private.
I'd move these fields into a single, static inner class. Its only instance might be available as a static final memorizing supplier here.
```
Suppliers.memoize(()->{
  XtextPluginImagesData[] result = new XtextPluginImagesData[1];
  if (Display.getCurrent() != null) { // display thread, init directly
    result[0] = new XtextPluginImagesData();
  } else {
    Display.syncExec(()->result[0] = new XtextPluginImagesData());
  }
  return result[0];
});
```

Reply:
@szarnekow Interesting!

I can try to apply it, but wouldn't something like changing this initialization:

```java
	/* package */
	private static final void initialize() {
		PLUGIN_REGISTRY = new ImageRegistry();
```

to something like this:

```java
	/* package */
	private static final void initialize() {
		Display display = Display.getCurrent();
		PLUGIN_REGISTRY = new ImageRegistry(display != null ? display : Display.getDefault());
```

be enough? Of course, disregarding my initial change.

Reply:
If in doubt, org.eclipse.jface.resource.JFaceResources.getResources(Display) requires the display thread in `toQuery.disposeExec`

Reply:
I see.
But then, something like this should work, shouldn't it? (untested)

```java
private static final void initialize() {
	if (Display.getCurrent() != null)
		PLUGIN_REGISTRY = new ImageRegistry();
	else
		Display.getDefault().syncExec(()->PLUGIN_REGISTRY = new ImageRegistry());
...
```

I'm just asking to see whether there's a simpler solution;

Reply:
We'd still need to ensure that no concurrent reads happen while the maps are being populated. So replacing the four fields with a single container object that's created atomically sounds like the best thing we can do here.

Might be that we can avoid the Supplier and do the synchronized null check manually. But I don't see much value in inlining the logic that Suppliers.memoize offers.

Reply:
@szarnekow I see: I didn't notice that `Suppliers.memoize` is thread-safe https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)

I'll work on your proposal ASAP

Reply:
Maybe add a null-guard in line 138 and other places where the PLUGIN_REGISTRY is dereferenced?

Reply:
So going back to my original change but with additional null guards in other places?